src/lib/graphqlTimeoutMiddleware.ts

```diff
@@ -50,18 +71,16 @@ export const graphqlTimeoutMiddleware = (defaultTimeoutInMS: number) => {
           reject(new GraphQLTimeoutError(`GraphQL Timeout Error: ${field} has timed out after waiting for ${timeoutInMS}ms`))
         }, timeoutInMS)
       }),
-      new Promise(async (resolve, reject) => {
-        try {
-          const result = await middlewareResolver(parent, args, context, info)
-          resolve(result)
-        }
-        catch (error) {
-          reject(error)
-        }
-        finally {
+      resolverResult.then(
```

Comment:
This is dense so I want to make sure I understand it: rather than resolving the middleware in a promise race, you do it above and keep a reference to `resolverResult`. If nothing is waiting on resolverResult (I'm still fuzzy on what exactly `isPendingPromise` is doing?) then go ahead and just return it.

However, if it has been `then`'d and something is waiting for it, race it against a promise that autorejects within a given timeout window. 

This is how we add timeouts to all resolvers, right, if the timeout value is > 0?

I'm going to try to understand the `isPendingPromise` a bit better before, but this looks clever and good.

Reply:
Yup, you got it 👌